From 1c238932748a3ad9a9c39e2d61ef74e80aa5dc2c Mon Sep 17 00:00:00 2001
From: Cristian Sovaiala <cristian.sovaiala@freescale.com>
Date: Thu, 14 Mar 2013 19:56:11 +0000
Subject: [PATCH 038/430] dpaa_eth: MTU checking is not necessary for IP
 Reassembled frames

Do not limit IP Reassembled frames to MTU size.

Change-Id: I8ec3ca5e23165d7ee6397935dff26c7a18903fef
Signed-off-by: Cristian Sovaiala <cristian.sovaiala@freescale.com>
Reviewed-on: http://git.am.freescale.net:8181/1280
Reviewed-by: Fleming Andrew-AFLEMING <AFLEMING@freescale.com>
Tested-by: Fleming Andrew-AFLEMING <AFLEMING@freescale.com>
[original patch is from QorIQ-SDK-V1.4-20130625-yocto]
Signed-off-by: Jiang Lu <lu.jiang@windriver.com>
---
 drivers/net/ethernet/freescale/dpa/dpaa_eth.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/drivers/net/ethernet/freescale/dpa/dpaa_eth.c b/drivers/net/ethernet/freescale/dpa/dpaa_eth.c
index e22e508..e216c99 100644
--- a/drivers/net/ethernet/freescale/dpa/dpaa_eth.c
+++ b/drivers/net/ethernet/freescale/dpa/dpaa_eth.c
@@ -2104,7 +2104,9 @@ shared_rx_dqrr(struct qman_portal *portal, struct qman_fq *fq,
 skb_copied:
 	skb->protocol = eth_type_trans(skb, net_dev);
 
-	if (unlikely(dpa_check_rx_mtu(skb, net_dev->mtu))) {
+	/* IP Reassembled frames are allowed to be larger than MTU */
+	if (unlikely(dpa_check_rx_mtu(skb, net_dev->mtu) &&
+		!(fd->status & FM_FD_IPR))) {
 		percpu_priv->stats.rx_dropped++;
 		dev_kfree_skb_any(skb);
 		goto out;
-- 
1.7.5.4

